File: canConstruct.py
'''
Wrtie a function 'canConstruct(target, wordBank)' that accepts a 
target string and an array of strings.

The function should return a boolean indicating whether or not the
'target' can be constructed by concatenating elements of the
'wordBank' array.

You may reuse  elements of the 'wordBank' as many times as needed.
'''

'''
Brute Force:
Time complexity: O(n^m *m)
Space complexity: O(m^2)

Memoized Implementation:
Time Complexity: O(n*m^2)
Space Complexity: O(m^2)
'''

def canConstruct(target, wordBank, memo = {}):
    if (target in memo):
        return memo[target]

    if (target == ""):
        return True
    
    for word in wordBank:
        if (target.find(word) == 0):
            suffix = target[len(word):]
            if (canConstruct(suffix, wordBank, memo) == True):
                memo[target] = True
                return True
    memo[target] = False
    return False


if __name__ == "__main__":
    print(canConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"])) # True
    print(canConstruct("skateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"])) # False
    print(canConstruct("enterapotentpot", ["a", "p", "ent", "enter", "ot", "o", "t"])) # True
    print(canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"])) # False

File: canConstruct.cpp
 #include <iostream>
 #include <unordered_map>
 #include<vector>

 bool canConstruct(std::string target, std::vector<std::string> wordBank,  std::unordered_map<std::string, bool> &memo){
    if(memo.find(target) != memo.end()) return memo[target]; 
    if(target == "") return true;

    for(std::string word : wordBank){
        int idx = target.find(word);
        if(idx == 0){
            std::string suffix = target.substr(idx + word.length());
            if (canConstruct(suffix, wordBank, memo) == true){
                memo[target] = true;
                return true;
            }
        }   
    }
    memo[target] = false;
    return false;
 }

 int main(){
    std::unordered_map<std::string, bool> memo;
    std::cout <<canConstruct("abcdef", {"ab", "abc", "cd", "def", "abcd"}, memo) << std::endl;
    std::cout <<canConstruct("skateboard", {"bo", "rd", "ate", "t", "ska", "sk", "boar"}, memo)<< std::endl;
    std::cout <<canConstruct("enterapotentpot", {"a", "p", "ent", "enter", "ot", "o", "t"}, memo)<< std::endl;
    std::cout <<canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeef", {"e", "ee", "eee", "eeee", "eeeee", "eeeeee"}, memo)<< std::endl;
    return 0;
 }